wgengine/magicsock/relaymanager.go

```diff
@@ -17,35 +25,150 @@ import (
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
-	di, ok := h.discoInfoByServerDisco[serverDisco]
+func (r *relayManager) discoInfo(serverDisco key.DiscoPublic) (_ *discoInfo, ok bool) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.initLocked()
+	di, ok := r.discoInfoByServerDisco[serverDisco]
 	return di, ok
 }
 
-func (h *relayManager) handleCallMeMaybeVia(dm *disco.CallMeMaybeVia) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
+func (r *relayManager) handleCallMeMaybeVia(dm *disco.CallMeMaybeVia) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.initLocked()
+	// TODO(jwhited): implement
+}
+
+func (r *relayManager) handleBindUDPRelayEndpointChallenge(dm *disco.BindUDPRelayEndpointChallenge, di *discoInfo, src netip.AddrPort, vni uint32) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.initLocked()
 	// TODO(jwhited): implement
 }
 
-func (h *relayManager) handleBindUDPRelayEndpointChallenge(dm *disco.BindUDPRelayEndpointChallenge, di *discoInfo, src netip.AddrPort, vni uint32) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
+// cancelOutstandingWork cancels any in-progress work for 'ep'.
+func (r *relayManager) cancelOutstandingWork(ep *endpoint) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	existing, ok := r.relaySetupWorkByEndpoint[ep]
+	if ok {
+		existing.cancel()
+		existing.wg.Wait()
+		delete(r.relaySetupWorkByEndpoint, ep)
+	}
+}
+
+// allocateAndHandshakeAllServers kicks off allocation and handshaking of relay
+// endpoints for 'ep' on all known relay servers, canceling any existing
+// in-progress work.
+func (r *relayManager) allocateAndHandshakeAllServers(ep *endpoint) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.initLocked()
+	existing, ok := r.relaySetupWorkByEndpoint[ep]
+	if ok {
+		existing.cancel()
+		existing.wg.Wait()
+		delete(r.relaySetupWorkByEndpoint, ep)
+	}
+	if len(r.serversByAddrPort) == 0 {
+		return
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	started := &relaySetupWork{ep: ep, cancel: cancel, wg: &sync.WaitGroup{}}
+	for k := range r.serversByAddrPort {
+		started.wg.Add(1)
+		go r.allocateAndHandshakeForServer(ctx, started.wg, ep, k)
+	}
+	r.relaySetupWorkByEndpoint[ep] = started
+
+	go func() {
+		started.wg.Wait()
+		started.cancel()
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		maybeCleanup, ok := r.relaySetupWorkByEndpoint[ep]
+		if ok && maybeCleanup == started {
+			// A subsequent call to allocateAndHandshakeAllServers may have raced to
+			// delete the associated key/value, so ensure the work we are
+			// cleaning up from the map is the same as the one we were waiting
+			// to finish.
+			delete(r.relaySetupWorkByEndpoint, ep)
+		}
+	}()
+}
+
+func (r *relayManager) handleNewServerEndpoint(ctx context.Context, wg *sync.WaitGroup, server netip.AddrPort, se udprelay.ServerEndpoint) {
 	// TODO(jwhited): implement
 }
+
+func (r *relayManager) allocateAndHandshakeForServer(ctx context.Context, wg *sync.WaitGroup, ep *endpoint, server netip.AddrPort) {
```

Comment:
:pick: Nitpick: The name of your methods implies that `server` should come before the endpoint. Can I suggest this for consistency between `handleNewServerEndpoint` and `allocateAndHandshakeForServer`? I find that I’m not smart enough to keep track of function signatures.

```suggestion
func (r *relayManager) handleNewServerEndpoint(ctx context.Context, wg *sync.WaitGroup, server netip.AddrPort, se udprelay.ServerEndpoint) {
	// TODO(jwhited): implement
}

func (r *relayManager) allocateAndHandshakeForServer(ctx context.Context, wg *sync.WaitGroup, server netip.AddrPort, ep *endpoint) {
```

Reply:
sure, updated